File: class23_homework/B-Georgi-Ivanov-09/Java/Plane.java
package Homework23;

public class Plane {
	private int [][] seats;
	private int emptySeasts;
	private int numOfRows;
	private int numOfSeatsPerRow;
	private int afterWhichSeatIsTheWalkway;
	
	Plane(int rows, int seatsPerRow, int afterWhichSeatIsTheWalkway){
		seats = new int[rows+1][seatsPerRow+1]; // Just so we start from the first seat of the first row rather than the 0 seat of the 0 row...
		setNumOfRows(rows);
		setNumOfSeatsPerRow(seatsPerRow);
		setEmptySeasts(rows*seatsPerRow);
		setAfterWhichSeatIsTheWalkway(afterWhichSeatIsTheWalkway);
		emptyPlane();
	}		
	
	public void flyPlane(){
		System.out.println("Flying away...");
	}
	
	public void fillPlane(){
		while(getEmptySeasts() > 0){
			addGroupToPlane();
		}
		System.out.println("Plane is full!");
	}
	
	public void addGroupToPlane(){
		Group newGroup = new Group();
		System.out.println("Created group of " + newGroup.getPeopleInGroup());
		findSeats(newGroup);
	}
	
	public void findSeats(Group newGroup){
		if(newGroup.getPeopleInGroup() == 1){
			addOnePerson(newGroup);
		}else if(newGroup.getPeopleInGroup() == 2 && getEmptySeasts() >= 2){
			addTwoPeople(newGroup);
		}else if(newGroup.getPeopleInGroup() == 3 && getEmptySeasts() >= 3){
			addThreePeople(newGroup);
		}else{
			System.out.println("Not enough seats!");
		}
	}
	
	public void addOnePerson(Group newGroup){
		for(int rowCounter = 1; rowCounter <= getNumOfRows(); rowCounter++){
			for(int seatsCounter = 1; seatsCounter <= getNumOfSeatsPerRow(); seatsCounter++){
				if(seats[rowCounter][seatsCounter] == 0){ // Seat is empty
					seats[rowCounter][seatsCounter] = 1; // 1 because the passenger is alone :\
					emptySeasts--;
					System.out.println("Passenger 1 seat is row: " + rowCounter + ", seat: " + seatsCounter);
					return; // Exiting method
				}
			}
		}
	}
	
	public void addTwoPeople(Group newGroup){
		for(int rowCounter = 1; rowCounter <= getNumOfRows(); rowCounter++){
			for(int seatsCounter = 1; seatsCounter <= getNumOfSeatsPerRow()-1; seatsCounter++){
				if(seats[rowCounter][seatsCounter] == 0 && seats[rowCounter][seatsCounter+1] == 0){ // If this seat and the next one are free
					if(!(seatsCounter == getAfterWhichSeatIsTheWalkway() && seatsCounter+1 == getAfterWhichSeatIsTheWalkway()+1)){ // If the seats are separated by the walkway
						seats[rowCounter][seatsCounter] = 2;
						seats[rowCounter][seatsCounter+1] = 2;
						emptySeasts -= 2;
						System.out.println("Passenger 2 seat is row: " + rowCounter + ", seat: " + seatsCounter);
						System.out.println("Passenger 2's friend's seat is row: " + rowCounter + ", seat: " + (seatsCounter+1));
						return; // Exiting method
					}
				}
			}
		}
	}
	
	// Trying to understand this method can cause dizziness... Proceed at your own risk!
	public void addThreePeople(Group newGroup){
		for(int rowCounter = 1; rowCounter <= getNumOfRows(); rowCounter++){
			for(int seatsCounter = 1; seatsCounter <= getNumOfSeatsPerRow()-2; seatsCounter++){
				if(seats[rowCounter][seatsCounter] == 0 && seats[rowCounter][seatsCounter+1] == 0 && seats[rowCounter][seatsCounter+2] == 0){
					if(!(seatsCounter == getAfterWhichSeatIsTheWalkway() && seatsCounter+1 == getAfterWhichSeatIsTheWalkway()+1 && seatsCounter+2 == getAfterWhichSeatIsTheWalkway()+2)){ // checks if:  3 | 3 3
					if(!(seatsCounter+1 == getAfterWhichSeatIsTheWalkway() && seatsCounter+2 == getAfterWhichSeatIsTheWalkway()+1)){ // checks if:  3 3 | 3
						seats[rowCounter][seatsCounter] = 3;
						seats[rowCounter][seatsCounter+1] = 3;
						seats[rowCounter][seatsCounter+2] = 3;
						emptySeasts -= 3;
						System.out.println("Passenger 3 seat is row: " + rowCounter + ", seat: " + seatsCounter);
						System.out.println("Passenger 3's friend's seat is row: " + rowCounter + ", seat: " + (seatsCounter+1));
						System.out.println("Passenger 3's other friend's seat is row: " + rowCounter + ", seat: " + (seatsCounter+2));
						return; // Exiting method
						}
					}
				}
			}
		}
	}
	
	void emptyPlane(){
		for(int rowCounter = 1; rowCounter <= getNumOfRows(); rowCounter++){
			for(int seatsCounter = 1; seatsCounter <= getNumOfSeatsPerRow(); seatsCounter++){
				seats[rowCounter][seatsCounter] = 0; // Filling array with 0. 0 means a seat is empty
			}
		}
	}

	public void printAllSeats(){
	   for(int rowCounter = 1; rowCounter <= getNumOfRows(); rowCounter++){
	      for(int seatsCounter = 1; seatsCounter <= getNumOfSeatsPerRow(); seatsCounter++){
	         System.out.printf("%5d ", seats[rowCounter][seatsCounter]);  
	      }
	      System.out.println();
	   }
	}
	
	// GETTERS AND SETTERS
	public int getEmptySeasts() {
		return emptySeasts;
	}

	public void setEmptySeasts(int emptySeasts) {
		this.emptySeasts = emptySeasts;
	}

	public int getNumOfSeatsPerRow() {
		return numOfSeatsPerRow;
	}

	public void setNumOfSeatsPerRow(int numOfSeatsPerRow) {
		this.numOfSeatsPerRow = numOfSeatsPerRow;
	}

	public int getNumOfRows() {
		return numOfRows;
	}

	public void setNumOfRows(int numOfRows) {
		this.numOfRows = numOfRows;
	}

	public int getAfterWhichSeatIsTheWalkway() {
		return afterWhichSeatIsTheWalkway;
	}

	public void setAfterWhichSeatIsTheWalkway(int afterWhichSeatIsTheWalkway) {
		this.afterWhichSeatIsTheWalkway = afterWhichSeatIsTheWalkway;
	}		
}

File: class23_homework/B-Georgi-Ivanov-09/Ruby/Plane.rb
require_relative "Group"

class Plane
	def initialize rows, seats_per_row, after_which_seat_is_the_walkway
		@seats = Array.new(rows+1) { Array.new(seats_per_row+1, 0) } # Creating array and filling it with 0. 0 means a seat is empty
		@num_of_rows = rows
		@num_of_seats_per_row = seats_per_row
		@after_which_seat_is_the_walkway = after_which_seat_is_the_walkway
		@empty_seats = rows * seats_per_row
	end
	
	def fly_plane
		puts "Flying away..."
	end
	
	def fill_plane
		while @empty_seats > 0
			add_group_to_plane
		end
		puts "Plane is full!"
	end
	
	def add_group_to_plane
		new_group = Group.new
		puts "Ceated group of #{new_group.get_people_in_group}"
		find_seats new_group
	end
	
	def find_seats new_group
		if new_group.get_people_in_group == 1
			add_one_person
		elsif new_group.get_people_in_group == 2 && @empty_seats >= 2
			add_two_people
		elsif new_group.get_people_in_group == 3 && @empty_seats >= 3
			add_three_people
		else
			puts "Not enough seats!"
		end
	end
	
	def add_one_person
		@seats.each_index do |row|
			next if row == 0 # Skiping the 0 row
			@seats[row].each_index do |seat|
				next if seat == 0 # Skiping the 0 seat
				if @seats[row][seat] == 0
					puts "New passenger on row = #{row}, seat = #{seat}"
					@seats[row][seat] = 1
					@empty_seats -= 1
					return
				end
			end
		end
	end
	
	def add_two_people
		@seats.each_index do |row|
			next if row == 0 # Skiping the 0 row
			@seats[row].each_index do |seat|
				next if seat == 0 # Skiping the 0 seat
				if @seats[row][seat] == 0 && @seats[row][seat+1] == 0
					if seat != @after_which_seat_is_the_walkway && seat+1 != @after_which_seat_is_the_walkway+1
						@seats[row][seat] = 2
						@seats[row][seat+1] = 2
						puts "Passenger 2 seat is row = #{row}, seat = #{seat}"
						puts "Passenger 2's friend's seat is row = #{row}, seat = #{seat+1}"
						@empty_seats -= 2
						return
					end
				end
			end
		end
	end
	
	def add_three_people
		@seats.each_index do |row|
			next if row == 0 # Skiping the 0 row
			@seats[row].each_index do |seat|
				next if seat == 0 # Skiping the 0 seat
				if @seats[row][seat] == 0 && @seats[row][seat+1] == 0 && @seats[row][seat+2] == 0
					if seat != @after_which_seat_is_the_walkway && seat+1 != @after_which_seat_is_the_walkway+1 && seat+2 != @after_which_seat_is_the_walkway+2 # checks if:  3 | 3 3
					if seat+1 != @after_which_seat_is_the_walkway && seat+2 != @after_which_seat_is_the_walkway+1 # checks if:  3 3 | 3
						@seats[row][seat] = 3
						@seats[row][seat+1] = 3
						@seats[row][seat+2] = 3
						puts "Passenger 3 seat is row = #{row}, seat = #{seat}"
						puts "Passenger 3's friend's seat is row = #{row}, seat = #{seat+1}"
						puts "Passenger 3's other friend's seat is row = #{row}, seat = #{seat+2}"
						@empty_seats -= 3
						return
					end
					end
				end
			end
		end
	end

	def print_all_seats
		@seats.each_index do |row|
			next if row == 0 # Skiping the 0 row
			@seats[row].each_index do |seat|
				next if seat == 0 # Skiping the 0 seat
				print "  #{@seats[row][seat]}"
			end
			puts
		end
	end
	
end